Code/Framework/AzCore/AzCore/DOM/DomBackend.h

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/DOM/DomVisitor.h>
+#include <AzCore/IO/GenericStreams.h>
+#include <AzCore/IO/Path/Path.h>
+#include <AzCore/std/containers/vector.h>
+#include <AzCore/std/smart_ptr/unique_ptr.h>
+
+namespace AZ::Dom
+{
+    //! Backends are registered centrally and used to transition DOM formats to and from a textual format.
+    class Backend
+    {
+    public:
+        virtual ~Backend() = default;
+
+        //! Attempt to read this format from the given stream into the target Visitor.
+        //! The base implementation reads the stream into memory and calls ReadFromString.
+        virtual Visitor::Result ReadFromStream(
+            AZ::IO::GenericStream* stream, Visitor& visitor, size_t maxSize = AZStd::numeric_limits<size_t>::max());
+        //! Attempt to read this format from a mutable string into the target Visitor. This enables some backends to
+        //! parse without making additional string allocations.
+        //! This string may be modified and read in place without being copied, so when calling this please ensure that:
+        //! - The string won't be deallocated until the visitor no longer needs the values and
+        //! - The string is safe to modify in place.
+        //! The base implementation simply calls ReadFromString.
+        virtual Visitor::Result ReadFromStringInPlace(AZStd::string& buffer, Visitor& visitor);
+        //! Attempt to read this format from an immutable buffer in memory into the target Visitor.
+        virtual Visitor::Result ReadFromString(AZStd::string_view buffer, Lifetime lifetime, Visitor& visitor) = 0;
```

Comment:
`ReadFromStream` invoking `ReadFromString` seems backwards to me.

Because of the implementation of `ReadFromStream` create a local AZStd::string, it requires dynamic memory allocations where it shouldn't be needed.
If anything `ReadFromString` should create an `AZ::IO::MemoryStream` from the string_view and then invoke `ReadFromStream`.

Or better yet keep the API simpler and only provide a `ReadFromStream` function.

Reply:
While these can be specialized if there's call to avoid dynamic memory allocations, my logic here was that doing a full read into memory then processing should be a relatively performant default path (and is in line with how we deserialize JSON with our existing util functions).

Reply:
I see that as a different scenario, because the reading of the JSON into a buffer is part of the internal JSON serialization functionality.
It doesn't prescribe a behavior on the JSON serialization user.

In this case the implementer is expected to take care of reading from a stream.

Reply:
I would be _really_ careful with using streams. Since there are file streams I've seen major performance headaches due to excessive micro-reads because users can directly plug a file stream into one of these calls. Imho using O3DE's current streams sets users up for unexpected performance overhead, the kind they may not even see themselves depending on their hardware. I don't see a use case for streams here either, since the preferred use is to load the entire document into a mutable buffer and do in place loading. I would recommend removing support for streams entirely and only reconsider them again if there are clear use cases for them and even then if those use cases are for tools or for runtime or both.